Add a reference doc for the `adUnit` object

rmloveland opened the pull request:
This addresses https://github.com/prebid/prebid.github.io/issues/555

rmloveland commented:
Hey @mjacobsonny would appreciate a review on this.  It's a first cut that can definitely be fleshed out more, but I think I've captured a fair number of the various properties, and at least it's better than what we have now. :-)

  It doesn't mention multiformat yet so the PR stays modular, but can easily be updated with multiformat info once that content is live.

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+| Name         | Scope    | Type                                  | Description                                                                                                                                                         |
+|--------------+----------+---------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `code`       | Required | String                                | Unique identifier you create and assign to this ad unit.  Used to set query string targeting on the ad. If using GPT, we recommend setting this to slot element ID. |
+| `sizes`      | Required | Array[Number] or Array[Array[Number]] | All sizes this ad unit can accept.  Examples: `[400, 600]`, `[[300, 250], [300, 600]]`.  For 1.0 and later, prefer one of the `mediaTypes.*.sizes` listed below.    |
```
Instead of "For 1.0 and later, prefer one of the `mediaTypes.*.sizes` listed below.", let's say "For 1.0 and later, best practice is to define size(s) within the appropriate `mediaTypes.banner|native|video` object"

rmloveland replied:
Updated in f3ffb3a

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+|--------------+----------+---------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `code`       | Required | String                                | Unique identifier you create and assign to this ad unit.  Used to set query string targeting on the ad. If using GPT, we recommend setting this to slot element ID. |
+| `sizes`      | Required | Array[Number] or Array[Array[Number]] | All sizes this ad unit can accept.  Examples: `[400, 600]`, `[[300, 250], [300, 600]]`.  For 1.0 and later, prefer one of the `mediaTypes.*.sizes` listed below.    |
+| `bids`       | Required | Array[Object]                         | Each bid represents a request to a bidder.  For a list of properties, see [Bids](#bids) below.                                                                      |
```
"Array of bid objects representing the demand partners and associated parameters for a given adUnit.  See [Bids](#bids) below".

rmloveland replied:
Updated in 73a556d

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+| `code`       | Required | String                                | Unique identifier you create and assign to this ad unit.  Used to set query string targeting on the ad. If using GPT, we recommend setting this to slot element ID. |
+| `sizes`      | Required | Array[Number] or Array[Array[Number]] | All sizes this ad unit can accept.  Examples: `[400, 600]`, `[[300, 250], [300, 600]]`.  For 1.0 and later, prefer one of the `mediaTypes.*.sizes` listed below.    |
+| `bids`       | Required | Array[Object]                         | Each bid represents a request to a bidder.  For a list of properties, see [Bids](#bids) below.                                                                      |
+| `mediaTypes` | Optional | Object                                | Defines the media type of the ad.  For a list of properties, see [Media Types](#mediatypes) below.                                                                  |
```
I assume we'll modify this wording later to refer to multiple media types?

rmloveland replied:
Updated the description to note that there can be one or more media types on a single ad unit in d898dcb

Happy to update that description if you have additional comments!

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+{: .table .table-bordered .table-striped }
+| Name     | Scope                                                        | Type   | Description                                                                                    |
+|----------+--------------------------------------------------------------+--------+------------------------------------------------------------------------------------------------|
+| `banner` | Required, unless either of the other properties are present. | Object | Defines properties of a banner ad.  For examples, see [`adUnit.mediaTypes.banner`](#banner).   |
```
"At least one of the 'banner', 'native' or 'video' objects is required."

rmloveland replied:
Updated in 0d91a4c

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+{: .table .table-bordered .table-striped }
+| Name          | Scope    | Type                                  | Description                                                                                                                       |
+|---------------+----------+---------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------|
+| `sizes`       | Required | Array[Number] or Array[Array[Number]] | All sizes this ad unit can accept.  Examples: `[400, 600]`, `[[300, 250], [300, 600]]`.                                           |
```
"mediaTypes.native.sizes" should not be a property, as a native ad does not have a size.  Sizes may be specified on certain assets.

rmloveland replied:
Ah, nice catch - thanks!

Removed in deadcb4

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+| Name          | Scope    | Type                                  | Description                                                                                                                       |
+|---------------+----------+---------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------|
+| `sizes`       | Required | Array[Number] or Array[Array[Number]] | All sizes this ad unit can accept.  Examples: `[400, 600]`, `[[300, 250], [300, 600]]`.                                           |
+| `type`        | Optional | String                                | Used as a shorthand, e.g., `type: 'image'` implies required fields `image`, `title`, `sponsoredBy`, `clickUrl`.                   |
```
Can we link to http://prebid.org/dev-docs/show-native-ads.html#pre-defined-native-types here?

rmloveland replied:
Great idea, added link in 3525e50

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+|---------------+----------+---------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------|
+| `sizes`       | Required | Array[Number] or Array[Array[Number]] | All sizes this ad unit can accept.  Examples: `[400, 600]`, `[[300, 250], [300, 600]]`.                                           |
+| `type`        | Optional | String                                | Used as a shorthand, e.g., `type: 'image'` implies required fields `image`, `title`, `sponsoredBy`, `clickUrl`.                   |
+| `title`       | Required | Object                                | The title of the ad, usually a call to action or a brand name.  For properties, see [`native.title`](#native-title).              |
```
The title is not the same as a call to action or brand name (each of which have dedicated assets).  Let's say something like "The title object is to be used for the title element of the Native ad."

rmloveland replied:
Updated in d371f5b

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+|---------------+----------+---------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------|
+| `sizes`       | Required | Array[Number] or Array[Array[Number]] | All sizes this ad unit can accept.  Examples: `[400, 600]`, `[[300, 250], [300, 600]]`.                                           |
+| `type`        | Optional | String                                | Used as a shorthand, e.g., `type: 'image'` implies required fields `image`, `title`, `sponsoredBy`, `clickUrl`.                   |
+| `title`       | Required | Object                                | The title of the ad, usually a call to action or a brand name.  For properties, see [`native.title`](#native-title).              |
```
None of the native assets (title, body, sponsoredBy, icon, image, clickUrl, cta) should be required in the mediaTypes.native object.

rmloveland replied:
Updated to mark all asset fields as optional in bccf932

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+| `sizes`       | Required | Array[Number] or Array[Array[Number]] | All sizes this ad unit can accept.  Examples: `[400, 600]`, `[[300, 250], [300, 600]]`.                                           |
+| `type`        | Optional | String                                | Used as a shorthand, e.g., `type: 'image'` implies required fields `image`, `title`, `sponsoredBy`, `clickUrl`.                   |
+| `title`       | Required | Object                                | The title of the ad, usually a call to action or a brand name.  For properties, see [`native.title`](#native-title).              |
+| `body`        | Required | Object                                | Text of the ad copy.  For properties, see [`native.body`](#native-body).                                                          |
```
"The body object is to be used for the body element of the Native ad."

rmloveland replied:
Updated in 900b2e4

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+| `body`        | Required | Object                                | Text of the ad copy.  For properties, see [`native.body`](#native-body).                                                          |
+| `sponsoredBy` | Required | String                                | The name of the brand associated with the ad.  For properties, see [`native.sponsoredBy`](#native-sponsoredby).                   |
+| `icon`        | Optional | Object                                | The brand icon that will appear with the ad.  For properties, see [`native.icon`](#native-icon).                                  |
+| `image`       | Optional | Object                                | A picture that is associated with the brand, or grabs the user's attention.  For properties, see [`native.image`](#native-image). |
```
"The image object is to be used for the main image of the Native ad."

rmloveland replied:
Updated in 27bc64b

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+| `sponsoredBy` | Required | String                                | The name of the brand associated with the ad.  For properties, see [`native.sponsoredBy`](#native-sponsoredby).                   |
+| `icon`        | Optional | Object                                | The brand icon that will appear with the ad.  For properties, see [`native.icon`](#native-icon).                                  |
+| `image`       | Optional | Object                                | A picture that is associated with the brand, or grabs the user's attention.  For properties, see [`native.image`](#native-image). |
+| `clickUrl`    | Optional | Object                                | Where the user will end up if they click the ad.  For properties, see [`native.clickUrl`](#native-clickurl).                      |
```
Are we sure that this is a property that publishers should be setting in the mediaTypes.native object?  All native bid responses should be required to provide a clickUrl regardless of the specification here.

rmloveland replied:
I think we should leave it in the ad unit docs until it's removed from the Prebid.js code.  Although it's a required part of the native bid response, for example it's currently listed as a part of the "image-type" native ad unit in [native.js](https://github.com/prebid/Prebid.js/blob/master/src/native.js#L24).

I can remove it now if you really want me to though, I don't hold the above opinion very strongly or think it makes much difference either way.  What do you think?

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+| `clickUrl`    | Optional | Object                                | Where the user will end up if they click the ad.  For properties, see [`native.clickUrl`](#native-clickurl).                      |
+| `cta`         | Optional | Object                                | *Call to Action* text, e.g., "Click here for more information".  For properties, see [`native.cta`](#native-cta).                 |
+
+<div class="alert alert-warning">
```
I kind of think that this warning seems a bit out of place in a page dedicated to the adUnit.  

In any section on Prebid Native Ad Validation, we should also mention that all native bid responses must define a landing page url.

rmloveland replied:
Removed the warning from this page in ca5181d

Updated the native ad validation section in 'Show Native Ads' to note that native bid responses must define a landing page URL in 083a247

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+{: .table .table-bordered .table-striped }
+| Name            | Scope    | Type                                  | Description                                                                                                                         |
+|-----------------+----------+---------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------|
+| `required`      | Required | Boolean                               | Whether this asset is required.                                                                                                     |
```
Is it true that the "required" property is required? (Same comment for all other asset objects below as well).

rmloveland replied:
> Is it true that the "required" property is required?

Looks like the answer is no, Prebid.js just filters the list to only required objects according to my reading of https://github.com/prebid/Prebid.js/blob/master/src/native.js#L105

Updated all mentions of the `required` property to note they are optional in dbc79b2

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+|-----------------+----------+---------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------|
+| `required`      | Required | Boolean                               | Whether this asset is required.                                                                                                     |
+| `sizes`         | Optional | Array[Number] or Array[Array[Number]] | All sizes this ad unit can accept.  Examples: `[400, 600]`, `[[300, 250], [300, 600]]`.                                             |
+| `aspect_ratios` | Optional | Array[Object]                         | Instead of `sizes`, you can define allowed aspect ratios.  For properties, see [`image.aspect_ratios`](#native-image-aspectratios). |
```
You say "instead", but can "aspect_ratios" be defined in addition to "sizes"?

rmloveland replied:
Looks like the answer is yes: https://github.com/prebid/Prebid.js/blob/master/test/spec/unit/core/adapterManager_spec.js#L836

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+See the table below for the list of properties in the `mediaTypes` object of the ad unit.  For example ad units showing the different media types, see the [Examples](#addAdUnits-Examples) below.
+
+{: .table .table-bordered .table-striped }
+| Name     | Scope                                                        | Type   | Description                                                                                    |
```
Let's order this "banner", "native", "video"

rmloveland replied:
Reordered in 59f410b

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+| Name       | Scope    | Type    | Description                                          |
+|------------+----------+---------+------------------------------------------------------|
+| `required` | Required | Boolean | Whether a title asset is required on this native ad. |
+| `len`      | Optional | Integer | Length of title text, in characters.                 |
```
"len" is meant to specify a maximum length, not an absolute length requirement.

rmloveland replied:
Thanks for the catch, updated in cecd7a4

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+
+<a name="native-clickUrl" />
+
+##### `adUnit.mediaTypes.native.clickUrl`
```
As mentioned above, "clickUrl" should always be implicitly required.

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+|------------+----------+---------+----------------------------------------------------------|
+| `required` | Required | Boolean | Whether a click URL asset is required on this native ad. |
+
+##### `adUnit.mediaTypes.native.body`
```
The "body" object should have a "len" property.

rmloveland replied:
Thanks, updated in be02c24

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+
+<a name="native-icon" />
+
+##### `adUnit.mediaTypes.native.icon`
```
If there's a clean way to consolidate native.icon with native.image, we should do so, since the properties below are identical.

rmloveland replied:
I share that desire, but I don't think we can implement it in a way that isn't a little confusing.  Right now the page has specific anchor links to object properties, which will be handy when linking from other pages, e.g.,

`<a name="adUnit.mediaTypes.native.icon" />`

If we consolidate I'm not sure how to make clear to someone clicking in from another page, who wants `adUnit.mediaTypes.native.icon.aspect_ratios` but gets the image version, that "oh these are the image docs but they're the same for icon".  We could add notes, except that we have to duplicate the note in a couple places so it's not much of a win dupe-wise IMO.

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+{: .table .table-bordered .table-striped }
+| Name            | Scope    | Type                                  | Description                                                                                                                       |
+|-----------------+----------+---------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------|
+| `required`      | Requird  | Boolean                               | Whether an icon asset is required on this ad.                                                                                     |
```
Typo on "Requird" (but should probably be "Optional" anyway).

rmloveland replied:
Updated all mentions of the required property to note they are optional in dbc79b2

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+{: .table .table-bordered .table-striped }
+| Name             | Scope    | Type                   | Description                                                                                                                                             |
+|------------------+----------+------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `context`        | Required | String                 | The video context, either `"instream"` or `"outstream"`.  Example: `context: "outstream"`                                                               |
```
At the moment, I think video.context is optional.

rmloveland replied:
Thanks, updated in 8b2c1c1

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+|------------------+----------+------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `context`        | Required | String                 | The video context, either `"instream"` or `"outstream"`.  Example: `context: "outstream"`                                                               |
+| `playerSize`     | Optional | Array[Integer,Integer] | The size (width, height) of the video player on the page, in pixels.  Example: `playerSize: [640, 480]`                                                 |
+| `mimes`          | Optional | Array[String]          | Required by OpenRTB when using [Prebid Server]({{site.baseurl}}/dev-docs/get-started-with-prebid-server.html).  Example: `mimes: ['video/mp4']`         |
```
The note about "Required by Prebid Server" should feature more prominently.  Let's change the scope to something like "Recommended.  Required by OpenRTB when using Prebid Server...". 

Description should be taken from OpenRTB spec:

Content MIME types supported (e.g., “video/x-ms-wmv”,
“video/mp4”)

rmloveland replied:
Updated in 53a6e3f

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+| `context`        | Required | String                 | The video context, either `"instream"` or `"outstream"`.  Example: `context: "outstream"`                                                               |
+| `playerSize`     | Optional | Array[Integer,Integer] | The size (width, height) of the video player on the page, in pixels.  Example: `playerSize: [640, 480]`                                                 |
+| `mimes`          | Optional | Array[String]          | Required by OpenRTB when using [Prebid Server]({{site.baseurl}}/dev-docs/get-started-with-prebid-server.html).  Example: `mimes: ['video/mp4']`         |
+| `protocols`      | Optional | Array[Integer]         | Required by OpenRTB when using [Prebid Server]({{site.baseurl}}/dev-docs/get-started-with-prebid-server.html).  Example: `protocols: [1,2,3,4,5,6,7,8]` |
```
Description should be taken from/ reference the OpenRTB spec:

"Array of supported video protocols"

rmloveland replied:
Updated in 53a6e3f

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+| `playerSize`     | Optional | Array[Integer,Integer] | The size (width, height) of the video player on the page, in pixels.  Example: `playerSize: [640, 480]`                                                 |
+| `mimes`          | Optional | Array[String]          | Required by OpenRTB when using [Prebid Server]({{site.baseurl}}/dev-docs/get-started-with-prebid-server.html).  Example: `mimes: ['video/mp4']`         |
+| `protocols`      | Optional | Array[Integer]         | Required by OpenRTB when using [Prebid Server]({{site.baseurl}}/dev-docs/get-started-with-prebid-server.html).  Example: `protocols: [1,2,3,4,5,6,7,8]` |
+| `playbackmethod` | Optional | Array[Integer]         | Required by OpenRTB when using [Prebid Server]({{site.baseurl}}/dev-docs/get-started-with-prebid-server.html).  Example: `playbackmethod: [2]`          |
```
Description should be taken from/ reference the OpenRTB spec.

rmloveland replied:
Updated in 53a6e3f

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,342 @@
+| `type`        | Optional | String                                | Used as a shorthand, e.g., `type: 'image'` implies required fields `image`, `title`, `sponsoredBy`, `clickUrl`.                   |
+| `title`       | Required | Object                                | The title of the ad, usually a call to action or a brand name.  For properties, see [`native.title`](#native-title).              |
+| `body`        | Required | Object                                | Text of the ad copy.  For properties, see [`native.body`](#native-body).                                                          |
+| `sponsoredBy` | Required | String                                | The name of the brand associated with the ad.  For properties, see [`native.sponsoredBy`](#native-sponsoredby).                   |
```
sponsoredBy is an Object, not a String.

rmloveland replied:
Nice catch, thanks!  Updated in a761885

rmloveland commented:
@mjacobsonny responding to some work items in your overall comment - 

> At the top, I think we should provide a high-level description of what an adUnit is.

Added a description in 34312ae, let me know what you think of it.

> The TOC looks a bit cluttered at the moment, but I don't necessarily have a better suggestion for how to fix it.

Not sure either.  My goal is for the TOC (and in-page anchor links) to match the "shape" of the object exactly so it's easier to jump to exactly what you need.  Unfortunately the visual of the TOC nested by property names is ... not great.

> I don't think that the section headings should be monospaced, and I'm not sure whether I like that each section header gets smaller as the objects get more nested.

Removed the monospacing from headers in f91563f.

I'm not a fan of the header text shrinking like that either.  I filed an issue to fix: https://github.com/prebid/prebid.github.io/issues/581

rmloveland commented:
@mjacobsonny do you have any additional feedback items on this PR I should address so we can merge?

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,414 @@
+    mediaTypes: {
+        video: {
+            context: "instream",
+            sizes: [640, 480],
```
Should be "playerSize" instead of "sizes".

rmloveland replied:
Addressed in 0333fd6

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,414 @@
+    mediaTypes: {
+        video: {
+            context: 'outstream',
+            sizes: [640, 480]
```
"playerSize" instead of "sizes"

rmloveland replied:
Addressed in 0333fd6

mjacobsonny commented on dev-docs/adunit-reference.md:
```diff
@@ -0,0 +1,414 @@
+            bids: [{
+                bidder: 'appnexus',
+                params: {
+                    placementId: 12340414,
```
Here, "placementId" is being passed as an int, but above it's being passed as a String.  Let's make sure we're consistent across examples.

rmloveland replied:
Thanks - updated in 42b9c7d to pass as string, since that's also what we show in the AN adapter example on the [Bidder params page](http://prebid.org/dev-docs/bidders.html#appnexus)

rmloveland commented:
Thanks again for the great review @mjacobsonny
